### PR TITLE
Break Apart Reducers: Act 2

### PR DESCRIPTION
**Pull request by jeremyckahn**
### What this PR does

This PR is part 2 of 2 for breaking the monolithic `reducers.js` file out into many files that each contain one reducer. It completes the work that was started in #280.

### How this change can be validated

Do a high-level smoke test of the game and ensure that nothing crashes. If something is broken, the game should crash immediately.

### Questions or concerns about this change

`src/game-logic/reducers/` now contains 125 files. 😳

### Additional information

Like #280, this is a pure refactor and no logic was changed. Though this PR is _huge_ (sorry...), it should hopefully be pretty skimmable.

### CODE REVIEW

File: src/game-logic/reducers/purchaseSmelter.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+import { showNotification } from './showNotification'
+import { updateLearnedRecipes } from './updateLearnedRecipes'
+
+/**
+ * @param {farmhand.state} state
+ * @param {number} smelterId
+ * @returns {farmhand.state}
+ */
+export const purchaseSmelter = (state, smelterId) => {
```
Comment (lstebner) [inline src/game-logic/reducers/purchaseSmelter.js:13]: 🎏

File: src/game-logic/reducers/purchaseStorageExpansion.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * @param {farmhand.state} state
+ * @returns {farmhand.state}
+ */
+export const purchaseStorageExpansion = state => {
+  const { money, inventoryLimit } = state
+  const storageUpgradeCost = getCostOfNextStorageExpansion(inventoryLimit)
+
+  if (money < storageUpgradeCost || inventoryLimit === -1) {
```
Comment (lstebner) [inline src/game-logic/reducers/purchaseStorageExpansion.js:12]: is -1 for ease of testing or a real feature?

File: src/game-logic/reducers/removeCowFromInventory.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import { addItemToInventory } from './addItemToInventory'
+
+import { changeCowBreedingPenResident } from './changeCowBreedingPenResident'
+
+/**
+ * @param {farmhand.state} state
+ * @param {farmhand.cow} cow
+ * @returns {farmhand.state}
+ */
+export const removeCowFromInventory = (state, cow) => {
```
Comment (lstebner) [inline src/game-logic/reducers/removeCowFromInventory.js:13]: 🎏

File: src/game-logic/reducers/removePeer.js [hunk 4]
```diff
@@ -0,0 +1,11 @@
+/**
+ * @param {farmhand.state} state
+ * @param {string} peerId The peer to remove
+ * @returns {farmhand.state}
+ */
+export const removePeer = (state, peerId) => {
```
Comment (lstebner) [inline src/game-logic/reducers/removePeer.js:6]: 🎏

File: src/game-logic/reducers/rotateNotificationLogs.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    info: [],
+    success: [],
+    warning: [],
+  }
+
+  newDayNotifications.forEach(({ message, severity }) =>
+    notifications[severity].push(message)
+  )
+
+  newDayNotifications.length &&
```
Comment (lstebner) [inline src/game-logic/reducers/rotateNotificationLogs.js:23]: can this be converted to an `if` block?

File: src/game-logic/reducers/upgradeTool.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import { TOOL_UPGRADED_NOTIFICATION } from '../../templates'
+
+import { showNotification } from './showNotification'
+import { makeRecipe } from './makeRecipe'
+
+/**
+ * @param {farmhand.state} state
+ * @param {farmhand.upgrade} upgrade
+ */
+export const upgradeTool = (state, upgrade) => {
```
Comment (lstebner) [inline src/game-logic/reducers/upgradeTool.js:11]: 🎏 (oops this one was probably one I should have added!)

File: src/game-logic/reducers/waterPlot.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+/**
+ * @param {farmhand.state} state
+ * @param {number} x
+ * @param {number} y
+ * @returns {farmhand.state}
+ */
+export const waterPlot = (state, x, y) => {
+  const plotContent = state.field[y][x]
+
+  if (!plotContent || getPlotContentType(plotContent) !== itemType.CROP) {
```
Comment (lstebner) [inline src/game-logic/reducers/waterPlot.js:15]: positive checks read a bit more easily to me here

```suggestion
const canBeWatered = plotContent && getPlotContentType(plotContent) === itemType.CROP

  if (canBeWatered === false) {
```

**Review by lstebner — APPROVED** [review]
I know none of this code is new, but since I haven't personally been in most of it I left comments around readility things I noticed. I definitely won't be offended if you don't take any of these suggestions, but I thought if I don't do it now there will never be another time I just go reading all these files so might as well! 

I also called out the ones that are missing specs like I did in the previous PR. I avoided commenting on test refactors though since I'd rather know they continue to pass if you do any refactoring to the actual logic files.

Based on the general fact that this was all code that you moved without modification I'm cool if you merge as is so I'm gonna approve even though I ended up leaving a lot of comments. Nice work getting through all of these!

File: src/game-logic/reducers/addCowToInventory.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+/**
+ * @param {farmhand.state} state
+ * @param {farmhand.cow} cow
+ * @returns {farmhand.state}
+ */
+export const addCowToInventory = (state, cow) => {
```
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/game-logic/reducers/addCowToInventory.js:7]: Good idea! I'm going to defer actually writing tests for now and just put in `TODO`s so these can be followed up with. This is done in https://github.com/jeremyckahn/farmhand/commit/bd4a4469350942dfbdd713f9f520ded38a1ffed5.

File: src/game-logic/reducers/upgradeTool.js (hunk 6 above)
Comment (jeremyckahn) [reply src/game-logic/reducers/upgradeTool.js:12]: Nobody's perfect! 😄

https://github.com/jeremyckahn/farmhand/commit/bd4a4469350942dfbdd713f9f520ded38a1ffed5

File: src/game-logic/reducers/removePeer.js (hunk 4 above)
Comment (jeremyckahn) [reply src/game-logic/reducers/removePeer.js:7]: https://github.com/jeremyckahn/farmhand/commit/bd4a4469350942dfbdd713f9f520ded38a1ffed5

File: src/game-logic/reducers/removeCowFromInventory.js (hunk 3 above)
Comment (jeremyckahn) [reply src/game-logic/reducers/removeCowFromInventory.js:14]: https://github.com/jeremyckahn/farmhand/commit/bd4a4469350942dfbdd713f9f520ded38a1ffed5

File: src/game-logic/reducers/purchaseSmelter.js (hunk 1 above)
Comment (jeremyckahn) [reply src/game-logic/reducers/purchaseSmelter.js:14]: https://github.com/jeremyckahn/farmhand/commit/bd4a4469350942dfbdd713f9f520ded38a1ffed5

File: src/game-logic/reducers/modifyCow.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
+/**
+ * @param {farmhand.state} state
+ * @param {string} cowId
+ * @param {Function(farmhand.cow)} fn Must return the modified cow or
+ * undefined.
+ * @returns {farmhand.state}
+ */
+export const modifyCow = (state, cowId, fn) => {
```
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/game-logic/reducers/modifyCow.js:9]: https://github.com/jeremyckahn/farmhand/commit/bd4a4469350942dfbdd713f9f520ded38a1ffed5

File: src/game-logic/reducers/addRevenue.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import { moneyTotal } from '../../utils'
+
+/**
+ * @param {farmhand.state} state
+ * @param {number} revenue
+ * @returns {farmhand.state}
+ */
+export const addRevenue = (state, revenueToAdd) => {
```
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/game-logic/reducers/addRevenue.js:9]: https://github.com/jeremyckahn/farmhand/commit/bd4a4469350942dfbdd713f9f520ded38a1ffed5

File: src/game-logic/reducers/addPeer.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
+/**
+ * @param {farmhand.state} state
+ * @param {string} peerId The peer to add
+ * @returns {farmhand.state}
+ */
+export const addPeer = (state, peerId) => {
```
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/game-logic/reducers/addPeer.js:7]: https://github.com/jeremyckahn/farmhand/commit/bd4a4469350942dfbdd713f9f520ded38a1ffed5

**Review by jeremyckahn — COMMENTED** [review]
Thanks for the great review @lstebner! I didn't have time to follow up with all of your comments this morning, but I'll update this PR as I work through them all!

File: src/game-logic/reducers/changeCowAutomaticHugState.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * @param {farmhand.state} state
+ * @param {farmhand.cow} cow
+ * @param {boolean} doAutomaticallyHug
+ * @returns {farmhand.state}
+ */
+export const changeCowAutomaticHugState = (state, cow, doAutomaticallyHug) => {
+  if (
+    (doAutomaticallyHug && !areHuggingMachinesInInventory(state.inventory)) ||
+    Boolean(cow.isUsingHuggingMachine) === doAutomaticallyHug
```
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/game-logic/reducers/changeCowAutomaticHugState.js]: I agree that this conditional logic is a little hard to follow, the suggested change didn't seem much clearer to me. I refactored this a different way in https://github.com/jeremyckahn/farmhand/commit/58014be85eecf42c2c399f1220828cf81dd3edeb. Let me know what you think!

File: src/game-logic/reducers/applyLoanInterest.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
+import { castToMoney, moneyTotal } from '../../utils'
+import { LOAN_INTEREST_RATE } from '../../constants'
+import { LOAN_BALANCE_NOTIFICATION } from '../../templates'
+
+/**
+ * @param {farmhand.state} state
+ * @returns {farmhand.state}
+ */
+export const applyLoanInterest = state => {
+  const loanBalance = moneyTotal(
```
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/game-logic/reducers/applyLoanInterest.js]: Yeah that's a good naming pattern. Done in https://github.com/jeremyckahn/farmhand/commit/1c5779330efd32f9cbde3bc47992b58533a18bb3!

File: src/game-logic/reducers/adjustLoan.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+ * @returns {farmhand.state}
+ */
+export const adjustLoan = (state, adjustmentAmount) => {
+  const loanBalance = moneyTotal(state.loanBalance, adjustmentAmount)
+  const money = moneyTotal(state.money, adjustmentAmount)
+
+  if (loanBalance === 0 && adjustmentAmount < 0) {
+    state = showNotification(state, LOAN_PAYOFF``, 'success')
+  } else if (adjustmentAmount > 0) {
+    // Player is taking out a new loan
```
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/game-logic/reducers/adjustLoan.js]: Done in https://github.com/jeremyckahn/farmhand/commit/af8f7784a24179897772c392c6004514c1d231e6! This is a good pattern and worth updating the codebase to use throughout. 🙂

File: src/game-logic/reducers/adjustLoan.js [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * @param {farmhand.state} state
+ * @param {number} adjustmentAmount This should be a negative number if the
+ * loan is being paid down, positive if a loan is being taken out.
+ * @returns {farmhand.state}
+ */
+export const adjustLoan = (state, adjustmentAmount) => {
+  const loanBalance = moneyTotal(state.loanBalance, adjustmentAmount)
+  const money = moneyTotal(state.money, adjustmentAmount)
```
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/game-logic/reducers/adjustLoan.js]: This is specifically tied to the `money` state property, so I went with `newMoney` in https://github.com/jeremyckahn/farmhand/commit/abc86521bfb2afab71b51a767eee0923cc7dc163.

File: src/game-logic/reducers/adjustLoan.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+  if (loanBalance === 0 && adjustmentAmount < 0) {
```
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/game-logic/reducers/adjustLoan.js]: :ok_hand: Done in https://github.com/jeremyckahn/farmhand/commit/abc86521bfb2afab71b51a767eee0923cc7dc163!

**Review by jeremyckahn — COMMENTED** [review]
Thanks for the excellent review @lstebner! Although I didn't make every suggested change, I did follow up with every thread. Please give the latest changes another review when you can and re-approve if it all looks good to you.

File: src/game-logic/reducers/generatePriceEvents.js [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+ * @param {farmhand.state} state
+ * @returns {farmhand.state}
+ */
+export const generatePriceEvents = state => {
+  const priceCrashes = { ...state.priceCrashes }
+  const priceSurges = { ...state.priceSurges }
+  let newDayNotifications = [...state.newDayNotifications]
+  let priceEvent
+
+  if (Math.random() < PRICE_EVENT_CHANCE) {
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/game-logic/reducers/generatePriceEvents.js:29]: hmm yeah that's a good point; i wonder if something in mining is sometimes rolling a "1" and that's where the <= came from.. dealing with this separate sounds smart to me

File: src/game-logic/reducers/purchaseCowPen.js [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+export const purchaseCowPen = (state, cowPenId) => {
+  const { money, purchasedCowPen } = state
+
+  if (purchasedCowPen >= cowPenId) {
+    return state
+  }
+
+  return {
+    purchasedCowPen: cowPenId,
+    money: moneyTotal(money, -PURCHASEABLE_COW_PENS.get(cowPenId).price),
```
[the remark this answers is not in the document]
Comment (lstebner) [reply src/game-logic/reducers/purchaseCowPen.js:18]: `adjustedBalance` or `adjustedIncome` maybe? mostly the word "adjust" or "adjusted" comes to mind.

**Review by lstebner — APPROVED** [review]
thanks for making the changes/adding TODOs! 🚀

File: src/game-logic/reducers/purchaseCowPen.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (jeremyckahn) [reply src/game-logic/reducers/purchaseCowPen.js:18]: I think something with "balance" might be good. Perhaps `computeBalance`?
